### PR TITLE
Implement Photon design for the viewer

### PR DESCRIPTION
**Pull request by sawanm9000**
- Flat icons
- Larger buttons make touch input easier
- No textures, unnecessary shadows and gradients - The overall UI is less taxing on the brain.
- No useless animations - Actions like hovering is registered fast - The overall UI now feels snappy.
- Lighter (as in kilobytes to load)
- New UI match with Firefox's UI.

Edit: Latest screenshot:
PDF js Dark

PDF js Light

Old look for comparison:
OLD PDF js

### CODE REVIEW

**Comment by Snuffleupagus** [comment]
Please note that this *may* apply here as well: https://github.com/mozilla/pdf.js/pull/10673#issuecomment-475874671, since the current UI was based on the only existing official design specification.

(In any case, please keep in mind that none of the `package.json` related changes belong in a patch like this.)

**Comment by Snuffleupagus** [comment]
Referencing issue #9702 (unfortunately there's not, as far as I know, an official Photon-compatible design specification for the PDF.js viewer).
 
*Please keep in mind that none of the `package.json` related changes belong in a patch like this.*

**Comment by Snuffleupagus** [comment]
Both of the files `package.json` and ` package-lock.json` must be completely removed from the patch, and please make sure to follow also see https://github.com/mozilla/pdf.js/wiki/Squashing-Commits.

**Comment by Snuffleupagus** [comment]
Both of the files `package.json` and ` package-lock.json` must be completely removed from the patch, and please make sure to follow https://github.com/mozilla/pdf.js/wiki/Squashing-Commits.

File: web/viewer.css
Comment (Snuffleupagus) [inline web/viewer.css]: Increasing these values will lead to *a lot* of wasted space for locales where the relevant strings are short, please revert this change!

File: web/viewer.css
Comment (Snuffleupagus) [inline web/viewer.css]: Similar to the comment above, please revert this since it leads to an incorrect/inconsistent width until the `select` has been localized.

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]
Both of the files `package.json` and `package-lock.json` must be completely removed from the patch, and please make sure to follow https://github.com/mozilla/pdf.js/wiki/Squashing-Commits.

---

> On the right-hand side of the Zoom dropdown, the border looks "weird".

This seems to be a more general problem, which actually applies to *every* single button on the main Toolbar.

Also, why does both the Zoom dropdown button, the Find next/previous buttons, and the Overlay buttons have a lighter background than every other button? This honestly looks quite out of place.

*Edit:* Furthermore, in some situations the pageNumber focus ring is now incorrectly red, see e.g. 
pageNumber_focus

*Edit2:* Also, assuming that the general design is deemed OK: I'm guessing that this is probably something which will benefit from being broken up into a couple of *separate* patches, to simplify testing *and* to reduce the risk of regressions!?

**Comment by utopianknight** [comment]
modern-look

The findbar now looks a lot like that of Firefox:
Firefox findbar

**Comment by timvandermeij** [comment]
I'm noticing the following things:

- When the viewer is loading, the page number indicator has a red border.
- Almost all icons appear to be off-center on hover (hover over the + in the toolbar for example and notice that the space on the right/bottom is larger than on the left/top).
- If I compare the design with the existing one, the background color of the toolbar stands out for me. The one from the existing design has a color that is almost equal to the background color of the viewer itself, whereas in this design the toolbar appears to be a lot lighter, which looks a bit strange to me.

Aside from the above, I do see improvements in this design over the existing one, but I'm definitely not a UI/UX designer. @shorlander Is this something you can chime in on? Is this something that would fit in with the current Firefox style or do you/others have different plans altogether for #9702 for example?

**Comment by utopianknight** [comment]
> If I compare the design with the existing one, the background color of the toolbar stands out for me. The one from the existing design has a color that is almost equal to the background color of the viewer itself, whereas in this design the toolbar appears to be a lot lighter, which looks a bit strange to me.

The colors have been updated. The scrollbars are now dark.

At this point, since there are no [guidelines ](https://design.firefox.com/photon/components/buttons.html)on some components, suggestions would be more helpful than criticisms.

**Comment by designakt** [comment]
@bwinton asked us to review the new UI before integrating into Firefox.

@aaronrbenson and I agree that this is a really nice match to other Firefox UI.
We only noticed a few things:

- **a few minor tweaks**
  - in Firefox we do not need the “open with” icon _(currently don't have that either)_
  - when clicking download, default should be safe file _(as is now)_
  - the hover area seems to be bigger then on other Firefox buttons - would be great to align
  - on mac the dropdown seems to have a glitch on top and bottom 
    CleanShot 2020-07-15 at 09 23 26@2x

- **a really nice to have**
  Ideally we would also have the toolbar align with Firefox toolbar style - **dark in dark mode, and light in light mode**. - With a line separating the pdf toolbar from what is above.

- **and a question**
  We like the new presentation mode icon, but were wondering what got us to use this over the previous symbol. Looking at it we noticed that we seem to mix "presentation mode" and "fullscreen" descriptions in this feature. **Do we know that presentation mode is better understood than fullscreen?**

For any follow-up questions please ping @aaronrbenson

**Comment by brendandahl** [comment]
@utopianknight I know this is very delayed, but do you think you'd have time to address the above comments and we'll get this merged in?

**Comment by utopianknight** [comment]
> on mac the dropdown seems to have a glitch on top and bottom

I don't have a Mac so I'm going to need help on this.

> Do we know that presentation mode is better understood than fullscreen?

Fullscreen button is for going full screen (it has its own icon). Presentation button is to switch to presentation mode. The reason why I changed the icon is so that it best communicate what the button actually **does**, which is to switch to presentation mode.

Question:
Pinging @aaronrbenson
This is how the dropdown button and a field look like respectively:
DropDownFieldButton

The reason why I changed the 'Zoom' dropdown button to make it look like a field is so that the user can type in the exact zoom level and fine-tune as they wish, while also be able to select zoom preset from a dropdown menu. This might be a solution to issue #1260. So which would be best? Should I make look like a dropdown button or leave it as it is?

@brendandahl
I'll work on it during the weekends.

**Comment by nt1m** [comment]
If this is blocked on the macOS select issue, I would recommend ignoring it for now, since this affects all `<select>` in general. I'm happy to look into fixing the root issue, though I'm first looking if there's an existing bug filed.

**Comment by nt1m** [comment]
I'm looking into it in: https://bugzilla.mozilla.org/show_bug.cgi?id=1653382

**Comment by utopianknight** [comment]
New update: Photon Design with dark and light theme and SVG icons.

[Compare](https://github.com/mozilla/pdf.js/compare/master...utopianknight:modern-look)

PDF js Dark

PDF js Light

High DPI preview:

PDF js Dark HiDPI

PDF js Light HiDPI

**Comment by bwinton** [comment]
@utopianknight It's pretty exciting to see the progress on this! Do you know when it will be ready for another UX review (or perhaps it's ready now)? 😃

**Comment by utopianknight** [comment]
@bwinton I don't know what the process is. I believe I'm finished with it but I'm also expecting your feedback so I can make further changes if necessary.

**Comment by aaronrbenson** [comment]
This is looking really good, thanks for your help @utopianknight! I can put some more time and thought into this next Monday and provide more feedback (if any) but for now I think these improvements look great. :shipit:

File: web/viewer.css
Comment (Snuffleupagus) [inline web/viewer.css:21]: These CSS variables should *not* be removed, please revert those changes.

File: web/viewer.css
Comment (Snuffleupagus) [inline web/viewer.css:21]: Unnecessary comment, please remove.

File: web/viewer.css
Comment (Snuffleupagus) [inline web/viewer.css:54]: The code below is pretty self-explanatory, hence this comment can be removed.

File: web/viewer.css
Comment (Snuffleupagus) [inline web/viewer.css:90]: Unnecessary comment, please remove.

File: web/viewer.css
Comment (Snuffleupagus) [inline web/viewer.css:1481]: Leftover comment at the end?

**Comment by timvandermeij** [comment]
The only thing I found is that the bookmark button has much more padding on hover than the other buttons, which looks a bit strange. This is pending UX/UI approval from Mozilla, but to me it looks very good!

File: web/viewer.css
Comment (Snuffleupagus) [inline web/viewer.css:1318]: The comment here is no longer aligned properly; and why even reformat this in the first place?

---

Given the overall size of this patch, please avoiding making unrelated changes.

File: web/viewer.css
Comment (Snuffleupagus) [inline web/viewer.css:1273]: Assuming that this no longer necessary, then it should be removed rather than be commented out.

File: web/viewer.css
Comment (Snuffleupagus) [inline web/viewer.css:1361]: Given that we (unfortunately) still need to support IE 11 in the default viewer, using `mask` is sadly not going to work (applies throughout the patch); note https://developer.mozilla.org/en-US/docs/Web/CSS/mask#Browser_compatibility

**Comment by utopianknight** [comment]
> the bookmark button has much more padding on hover than the other buttons

Fixed.

File: web/viewer.css
Comment (Snuffleupagus) [inline web/viewer.css:25]: Can you please use a *consistent* colour format throughout the patch, rather than mixing formats? 

There's at least four different formats used below, e.g. `#0c0c0d;`, `#5C5C61;`, `rgb(255, 74, 74);`, and `rgba(245, 246, 247, 1);`; please note PR #11447.

File: web/viewer.css
Comment (Snuffleupagus) [inline web/viewer.css:362]: These changes also appear unrelated to the patch, please revert.

File: web/viewer.css
Comment (Snuffleupagus) [inline web/viewer.css:288]: Leftover comment?

File: web/viewer.css
Comment (Snuffleupagus) [inline web/viewer.css:235]: This *appears* to be more unrelated re-formatting; please revert if there's no functional changes here.

File: web/viewer.css
Comment (Snuffleupagus) [inline web/viewer.css:261]: These changes look unrelated to the patch, and the added indentation makes it more difficult to read the rules below; please revert this.

File: web/viewer.css
Comment (Snuffleupagus) [inline web/viewer.css:267]: Same as the above.

File: web/viewer.css
Comment (Snuffleupagus) [inline web/viewer.css:384]: Looks like unnecessary reformatting.

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]

**Comment by fgilio** [comment]
This looks really good 😍
Thanks @utopianknight!

File: web/images/loading-dark.svg
Comment (Snuffleupagus) [inline web/images/loading-dark.svg:2]: Please note that the old `loading-small.png` files where *animated* PNG images, whereas this new SVG (and `loading.svg` as well) appear to be just stationary images.

File: web/viewer.css
Comment (Snuffleupagus) [inline web/viewer.css:91]: Nit: These values look identical to the ones above, is it necessary to redeclare them here?

**Comment by brendandahl** [comment]
@Snuffleupagus is this good to go?

**Comment by Snuffleupagus** [comment]
Unfortunately IE 11 is still causing trouble here, since the toolbar icons are not visible, please see

ie

**Comment by utopianknight** [comment]
It's seems to be working fine on my version of internet explorer.

Internetexplorer11

**Comment by Snuffleupagus** [comment]
The screen-shot in https://github.com/mozilla/pdf.js/pull/11077#issuecomment-673010632 was taken on Windows 7.

Furthermore the build logs contains the following output, which also needs fixing:
```
### Creating generic viewer
[17:51:14] gulp-postcss: viewer.css
postcss-css-variables: /home/ubuntu/pdfjs/botio-files-pdfjs/private/89bc36956c320d7/viewer.css:546:3: variable --scrollbar-color is undefined and used without a fallback
postcss-css-variables: /home/ubuntu/pdfjs/botio-files-pdfjs/private/89bc36956c320d7/viewer.css:546:3: variable --scrollbar-bg-color is undefined and used without a fallback
postcss-css-variables: /home/ubuntu/pdfjs/botio-files-pdfjs/private/89bc36956c320d7/viewer.css:546:3: variable --scrollbar-color is undefined and used without a fallback
postcss-css-variables: /home/ubuntu/pdfjs/botio-files-pdfjs/private/89bc36956c320d7/viewer.css:546:3: variable --scrollbar-bg-color is undefined and used without a fallback
postcss-css-variables: /home/ubuntu/pdfjs/botio-files-pdfjs/private/89bc36956c320d7/viewer.css:546:3: variable --scrollbar-color is undefined and used without a fallback
postcss-css-variables: /home/ubuntu/pdfjs/botio-files-pdfjs/private/89bc36956c320d7/viewer.css:546:3: variable --scrollbar-bg-color is undefined and used without a fallback
postcss-css-variables: /home/ubuntu/pdfjs/botio-files-pdfjs/private/89bc36956c320d7/viewer.css:546:3: variable --scrollbar-color is undefined and used without a fallback
postcss-css-variables: /home/ubuntu/pdfjs/botio-files-pdfjs/private/89bc36956c320d7/viewer.css:546:3: variable --scrollbar-bg-color is undefined and used without a fallback
```

**Comment by brendandahl** [comment]
I tried it out here on win10 IE11 and looks fine to me too.

**Review by Snuffleupagus — COMMENTED** [review]
Please note the build warnings mentioned in https://github.com/mozilla/pdf.js/pull/11077#issuecomment-674192940

Furthermore, there appears to be a merge-commit present in the PR, please make sure that you remove it; https://github.com/mozilla/pdf.js/wiki/Squashing-Commits may be helpful.

**Comment by Snuffleupagus** [comment]
From a support perspective, it's going to be *extremely* problematic if the viewer is broken in *some* IE 11 versions (maybe even related to the OS version).
Hence I think that there's basically two possible options here, either:
 - Attempt to fix this the `toolbar` icons, such that all the viewer works in all IE 11 versions.
 - Create a *new* PDF.js release first, and officially announce that it's the last one with support for IE 11 (after which this PR can land).

@brendandahl How should we proceed here?

**Comment by utopianknight** [comment]
I believe the following issue was fixed:

>       * on mac the dropdown seems to have a glitch on top and bottom
>         CleanShot 2020-07-15 at 09 23 26@2x

Testers needed.

**Comment by brendandahl** [comment]
Ideally, I'd like to not drop support just because of a UI refresh.
@Snuffleupagus what version of IE11 is on Win7? 

image

**Comment by utopianknight** [comment]
@Snuffleupagus Can you please use your developer tools and tell me what this is?

90047405-262dcc80-dcd2-11ea-81d4-952cdb211167

The only way for me to fix this is to use the same IE 11 version that you have. So I'd like to know this as well.

**Comment by utopianknight** [comment]
@brendandahl According to [this](https://docs.microsoft.com/en-us/lifecycle/faq/internet-explorer-microsoft-edge) and [this](https://support.microsoft.com/en-us/help/13853/windows-lifecycle-fact-sheet), Microsoft dropped support for Windows 7's Internet Explorer a while ago. Why are we still supporting it?

**Comment by Snuffleupagus** [comment]
> Ideally, I'd like to not drop support just because of a UI refresh.

Honestly though, it wouldn't *only* be because of the UI changes. The FAQ has listed IE 11 support as *limited* for well over a year now, see https://github.com/mozilla/pdf.js/wiki/Frequently-Asked-Questions#faq-support and issue #11211, and generally speaking supporting IE 11 is causing (and has historically caused) a fair amount of grief/problems/limitations over the years.

> @Snuffleupagus what version of IE11 is on Win7?

I cannot be sure that it's the *latest* available one, since I'm having trouble updating IE 11, however my currently installed version is:
ie11-version

---

> Can you please use your developer tools and tell me what this is?

Nothing special, as far as I can tell, it appears to simply be *part* of an SVG image.

Interestingly though it appears that all of the toolbar icons do load, however they're not scaled or positioned correctly; note what happens when I try disabling a randomly chosen `position` rule:
position

**Comment by Snuffleupagus** [comment]
Testing in Firefox, I noticed to additional things:
 - It appears that the updated `web/images/shadow.png` file becomes effectively *invisible* in `@media (prefers-color-scheme: dark)` mode; is it perhaps necessary to have two different versions of this image resource as well?
 - Also, if you're removing the `web/images/texture.png` file then please remove this line as well https://github.com/mozilla/pdf.js/blob/192afb87e34e1f7c30ad97e98599af643717eb9d/gulpfile.js#L772

**Comment by utopianknight** [comment]
Please note that prefers-color-scheme does not seem to work on the Developer Edition of Firefox without an extension. Until this is fixed, should the light theme be the default or the dark theme?

>It appears that the updated web/images/shadow.png file becomes effectively invisible

It is still visible to me. If you quickly toggle its visibility, you'll notice that something looks off when it's absent. The reason why I don't want to make it too dark is so that PDF files with dark pages don't appear too fuzzy on the edges. If you're asking me to use CSS shadow instead of border-image, please know that this could introduce performance issues.

**Comment by Snuffleupagus** [comment]
> It is still visible to me.

Really, even with the dark-mode enabled?
It obviously seems fine in the (default) light-mode; even looking at your screen-shots in https://github.com/mozilla/pdf.js/pull/11077#issue-308369353, `web/images/shadow.png` is not really discernible in the dark-mode (at least to my eye).

> If you're asking me to use CSS shadow instead of border-image, please know that this could introduce performance issues.

Of course not, I was around when the image was added (and the old CSS-only solution removed) for performance reasons :-)

**Comment by utopianknight** [comment]
This is an optical illusion. The shadow will always look invisible when the page and background are contrasting. Here's how it looks like with dark page:

Annotation 2020-08-17 163206

The only to fix it is to get the page's color and generate a shadow for each one of them. It's possible. Here's an example: https://github.com/robinloeffel/cosha

**Comment by timvandermeij** [comment]
For IE11, another scenario would be to do a major version bump for the next release and drop support completely. Not only do I agree with https://github.com/mozilla/pdf.js/pull/11077#issuecomment-674736476 but it would also simplify landing this patch so we don't have to waste time on IE11 anymore.

I think a major version bump would also be good given that many big features landed that received the most upvotes compared to other issues: AcroForms support, layers support, zoom fixes for OSX and TypeScript definitions. That's quite a bit and in my opinion deserves a major version bump.

What do you think @brendandahl?

**Comment by utopianknight** [comment]
@timvandermeij It's fixed. Please call /botio-linux

**Comment by brendandahl** [comment]
> For IE11, another scenario would be to do a major version bump for the next release (3.x) and drop support completely. Not only do I agree with [#11077 (comment)](https://github.com/mozilla/pdf.js/pull/11077#issuecomment-674736476) but it would also simplify landing this patch so we don't have to waste time on IE11 anymore. If someone _really_ needs IE11 support, the 2.x versions can still be used.
> 
> I think a major version bump would also be good given that many big features landed that received the most upvotes compared to other issues: AcroForms support, layers support, zoom fixes for OSX and TypeScript definitions. That's quite a bit and in my opinion deserves a major version bump.
> 
> What do you think @brendandahl?

I'm fine with the approach of dropping IE11 support. I'd like to stick to semantic versioning though, so I don't think we should bump the major version number since we haven't broken/deprecated any core API's.

@timvandermeij if we're fine with breaking IE11, is this good to go? I'd like to get it in for Firefox 81.

**Comment by timvandermeij** [comment]
If @utopianknight can fix this with the version information provided by @Snuffleupagus I'd prefer that, but looking at https://techcommunity.microsoft.com/t5/microsoft-365-blog/microsoft-365-apps-say-farewell-to-internet-explorer-11-and/ba-p/1591666 I'm fine with dropping IE11 support for the next version without a major version bump, as long as we make this clear in the commit message and during the release.

I haven't done an in-depth review of this patch, so I'll leave it to you and/or @Snuffleupagus to merge this.

/botio-linux preview

File: web/viewer.css
Comment (Snuffleupagus) [inline web/viewer.css:830]: Did you actually mean `--button-hover-color` here!?

Given that e.g. `.splitToolbarButton > .toolbarButton:hover` uses that css variable, the effect of *this* line is that different toolbar buttons end up looking slightly different when hovered.
Is this really intended, since it feels a bit inconsistent?

File: web/viewer.css
Comment (Snuffleupagus) [inline web/viewer.css:857]: I think that you'll need to add `max-width: 16px;` here as well, since otherwise the icon isn't visible on all IE 11 versions.

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]
The latest preview now *almost* works in IE 11, on Windows 7, note the screen-shot below; I've added an inline suggestion that should address the last missing icon.

Furthermore (and unrelated to IE), I've also noticed a slight inconsistency regarding the `.toolbarButton` hover effects, see the inline comment.

After the above have been fixed, I *believe* that this should now be good to land as-is even with regards to IE 11 on Windows 7 :-)

---
image

**Comment by utopianknight** [comment]
Alright. I'm done. I believe patch is ready to merge. Green light from me.

**Comment by Snuffleupagus** [comment]
The following is a, possibly incomplete, list of issues/bugs that's fixed by this PR:
 - #9702
 - #11572
 - #3528 (to the extent that we're going to "fix" it, since this patch should also *significantly* simplify UI customization)
 - [bug 1410598](https://bugzilla.mozilla.org/show_bug.cgi?id=1410598)
 - [bug 738947](https://bugzilla.mozilla.org/show_bug.cgi?id=738947) (that's essentially a WONTFIX at this point, since we're now aligning with the Photon design)

**Review by timvandermeij — APPROVED** [review]

**Comment by timvandermeij** [comment]
Looks good to me now. Thank you for all your work!

**Comment by szaimen** [comment]
🎉

**Comment by utopianknight** [comment]
Awesome! Great job everyone.

**Comment by natewind** [comment]
Wow, what an exceptionally ugly look! It’s like a textbook example of an unnecessary redesign, considering how perfect the old look was. Disgusting. @sawanm9000, you are the problem with modern design.

**Pull request merged**
